document.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// AddH2Header appends an H2 (##) header line to the document.
+func (doc *DocumentBuilder) AddH2Header(header string) error {
+	_, err := doc.body.WriteString("## ")
+	if err != nil {
+		return errors.Wrap(err, "Error writing to document")
+	}
+
+	err = doc.AddLine(header)
+	return err
+}
+
+// AddH3Header appends an H3 (###) header line to the document.
+func (doc *DocumentBuilder) AddH3Header(header string) error {
+	_, err := doc.body.WriteString("### ")
+	if err != nil {
+		return errors.Wrap(err, "Error writing header line to document")
+	}
+
+	err = doc.AddLine(header)
+	return err
+}
+
+// AddQuote appends a quote line to the document.
+func (doc *DocumentBuilder) AddQuote(header string) error {
+	_, err := doc.body.WriteString("> ")
+	if err != nil {
+		return errors.Wrap(err, "Error writing quote to document")
+	}
+
+	err = doc.AddLine(header)
+	return err
+}
+
+// AddBullet appends an unordered list item to the document.
+func (doc *DocumentBuilder) AddBullet(header string) error {
+	_, err := doc.body.WriteString("* ")
+	if err != nil {
+		return errors.Wrap(err, "Error writing bullet to document")
+	}
+
+	err = doc.AddLine(header)
+	return err
+}
+
+// ToggleFormatting appends a toggle formatting line to the document.
+func (doc *DocumentBuilder) ToggleFormatting() error {
+	return doc.AddLine("```")
+}
+
+// AddLink appends an aliased link line to the document.
+func (doc *DocumentBuilder) AddLink(url string, title string) error {
+	_, err := doc.body.WriteString("=> ")
+	if err != nil {
+		return errors.Wrap(err, "Error writing link to document")
+	}
+	_, err = doc.body.WriteString(url)
+	if err != nil {
+		return errors.Wrap(err, "Error writing link to document")
+	}
+	_, err = doc.body.WriteString("\t")
+	if err != nil {
+		return errors.Wrap(err, "Error writing link to document")
+	}
+	// AddLine to ensure there is a newline
+	err = doc.AddLine(title)
+	return err
+}
+
+// AddRawLink appends a link line to the document.
+func (doc *DocumentBuilder) AddRawLink(url string) error {
+	_, err := doc.body.WriteString("=> ")
+	if err != nil {
+		return errors.Wrap(err, "Error writing raw link to document")
+	}
+	err = doc.AddLine(url)
+	return err
+}
+
+// Build builds the document into a serialized byte slice.
+func (doc *DocumentBuilder) Build() ([]byte, error) {
+	buf := bytes.Buffer{}
+
+	// Write header
+	_, err := buf.WriteString(doc.header)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error building document header")
+	}
+	if !strings.HasSuffix(doc.header, "\n") {
+		_, err = buf.WriteString("\n")
+		if err != nil {
+			return nil, errors.Wrap(err, "Error building document header")
+		}
+	}
+
+	// Write body
+	_, err = buf.WriteString(doc.body.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "Error building document body")
+	}
+
+	// Write footer
+	_, err = buf.WriteString(doc.footer)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error building document footer")
+	}
+
+	return buf.Bytes(), nil
+}
```

Comment:
There is a lot going on here, but not as much as it looks. I believe we need to ensure there are newlines in between the header and body, which makes this a bit longer than it otherwise would. Also, this error handling feels a bit more verbose than normal. I'm a bit rusty with my Go though so perhaps there is a more concise way to handle these.